contributors/design-proposals/controller_history.md

```diff
@@ -0,0 +1,439 @@
+for this purpose.
+
+``` golang
+// ControllerRevision implements an immutable snapshot of state data. Clients 
+// are responsible for serializing and deserializing the objects that contain 
+// their internal state. 
+// Once a ControllerRevision has been successfully created, it can not be updated. 
+// The API Server will fail validation of all requests that attempt to mutate 
+// the Data field. ControllerRevisions may, however, be deleted.
+type ControllerRevision struct {
+    metav1.TypeMeta
+    // +optional
+    metav1.ObjectMeta
+    // Data contains the serialized state.
+    Data runtime.RawExtension
+    // Revision indicates the revision of the state represented by Data.
+    Revision int64
+}
+```
+
+## API Server
+The API Server must support the creation and deletion of ControllerRevision 
+objects. As we have no mechanism for declarative immutability, the API server 
+must fail any update request that updates the `.Data` field of a 
+ControllerRevision Object.
+
+## Controllers 
+This section is presented as a generalization of how an arbitrary controller 
+can use ControllerRevision to persist a history of revisions to its 
+specification type Objects. The technique is applicable, without loss of 
+generality, to the existing Kuberentes controllers that have Pod as a generated 
+type.
+
+When a controller detects a revision to the target Object state of a 
+specification type Object it will do the following.
+
+1. The controller will [create a snapshot](#version-snapshot-creation) of the 
+current target Object state.
+1. The controller will [reconstruct the history](#history-reconstruction) of 
+revisions to the Object's target Object state.
+1. The controller will test the current target Object state for 
+[equivalence](#version-equivalence) with all other versions in the Object's 
+revision history.
+    - If the current version is semantically equivalent to its immediate 
+    predecessor no update to the Object's target state has been performed.
+    - If the current version is equivalent to a version prior to its immediate 
+    predecessor, this indicates a rollback.
+    - If the current version is not equivalent to any prior version, this 
+    indicates an update or a roll forward.
+    - Controllers should use their status objects for book keeping with respect 
+    to current and prior revisions.
+1. The controller will 
+[reconcile its generated Objects](#target-object-state-reconciliation) 
+with the new target Object state.
+1. The controller will [maintain the length of its history](#history-maintenance) 
+to be less than the configured limit.
+
+### Version Snapshot Creation
+To take a snapshot of the target Object state contained in a specification type
+Object, a controller will do the following.
+
+1. The controller will serialize all the Object's target object state and store
+the serialized representation in the ControllerRevision's `.Data`.
```

Comment:
I left this general on purpose so that client code, StatefulSet controller, DaemonSet controller, operators, etc, is not constrained. There are examples of what the target Object state is above, (i.e it is called out that for ReplicaSet, DaemonSet, and StatefulSet this include the PodTemplateSpec, and that this includes the VolumeClaimsTemplate for StatefulSet). I'd prefer to defer the specifics of capturing the state to the StatefulSet update proposal and DaemonSet update proposal.